caffe2/opt/converter.cc

```diff
@@ -420,12 +442,88 @@ caffe2::NetDef convertToCaffe2Proto(repr::NNModule &m) {
   return convertToCaffe2Proto(m, predictNet);
 }
 
+bool shouldBeInplace(repr::NNGraph::NodeRef node) {
+  CAFFE_ENFORCE(repr::nn::is<repr::NeuralNetOperator>(node));
+  auto op = repr::nn::get<repr::NeuralNetOperator>(node);
+  using NNKind = repr::NeuralNetOperator::NNKind;
+  switch (op->getKind()) {
+    case NNKind::Relu:
+      return true;
+    default:
+      return false;
+  }
+}
+
+void addVersions(repr::NNModule* nn) {
+  std::unordered_map<std::string, uint64_t> versionMap;
+  for (auto& node : nn->dataFlow.getMutableNodes()) {
+    if (repr::nn::is<repr::NeuralNetData>(node)) {
+      auto d = repr::nn::get<repr::NeuralNetData>(node);
+      d->setVersion(versionMap[d->getName()]++);
+    }
+  }
+  for (auto& node_pair :
+       repr::nn::dataIterator<repr::NeuralNetOperator>(nn->dataFlow)) {
+    repr::NNGraph::NodeRef node;
+    repr::NeuralNetOperator* nnOp;
+    std::tie(nnOp, node) = node_pair;
+    // If we have an in-place node, all inputs/outputs should be versioned
+    // accordingly.
+    if (shouldBeInplace(node)) {
+      for (auto& input : repr::nn::getInputs(node)) {
+        for (auto& output : repr::nn::getInputs(node)) {
```

Comment:
`O(n^2)` is not ideal. :)

Reply:
yea, I agree -- this really should be something like a map so that we get fast query + order.